src/jimgw/transforms.py

```diff
@@ -0,0 +1,211 @@
+from abc import ABC, abstractmethod
+from typing import Callable
+
+import jax
+import jax.numpy as jnp
+from chex import assert_rank
+from jaxtyping import Float
+
+
+class Transform(ABC):
+    """
+    Base class for transform.
+
+    The idea of transform should be used on distribtuion,
+    """
+
+    name_mapping: tuple[list[str], list[str]]
+    transform_func: Callable[[dict[str, Float]], dict[str, Float]]
+
+    def __init__(
+        self,
+        name_mapping: tuple[list[str], list[str]],
+    ):
+        self.name_mapping = name_mapping
+
+    def __call__(self, x: dict[str, Float]) -> tuple[dict[str, Float], Float]:
+        return self.transform(x)
+
+    @abstractmethod
+    def transform(self, x: dict[str, Float]) -> tuple[dict[str, Float], Float]:
+        """
+        Transform the input x to transformed coordinate y and return the log Jacobian determinant.
+        This only works if the transform is a N -> N transform.
+
+        Parameters
+        ----------
+        x : dict[str, Float]
+                The input dictionary.
+
+        Returns
+        -------
+        y : dict[str, Float]
+                The transformed dictionary.
+        log_det : Float
+                The log Jacobian determinant.
+        """
+        raise NotImplementedError
+
+    @abstractmethod
+    def forward(self, x: dict[str, Float]) -> dict[str, Float]:
+        """
+        Push forward the input x to transformed coordinate y.
+
+        Parameters
+        ----------
+        x : dict[str, Float]
+                The input dictionary.
+
+        Returns
+        -------
+        y : dict[str, Float]
+                The transformed dictionary.
+        """
+        raise NotImplementedError
+
+    def propagate_name(self, x: list[str]) -> list[str]:
+        input_set = set(x)
+        from_set = set(self.name_mapping[0])
+        to_set = set(self.name_mapping[1])
+        return list(input_set - from_set | to_set)
+
+
+class UnivariateTransform(Transform):
+
+    def __init__(
+        self,
+        name_mapping: tuple[list[str], list[str]],
+    ):
+        super().__init__(name_mapping)
+
+    def transform(self, x: dict[str, Float]) -> tuple[dict[str, Float], Float]:
+        input_params = x.pop(self.name_mapping[0][0])
+        assert_rank(input_params, 0)
+        output_params = self.transform_func(input_params)
+        jacobian = jax.jacfwd(self.transform_func)(input_params)
+        x[self.name_mapping[1][0]] = output_params
+        return x, jnp.log(jacobian)
+
+    def forward(self, x: dict[str, Float]) -> dict[str, Float]:
+        input_params = x.pop(self.name_mapping[0][0])
```

Comment:
Moreover, how would it interface with the flowMC NF proposal? I would expect that to only be proposing new points in the transformed space. Does it mean that
1. an inverse transform is always needed, either for data product or likelihood calculation?

Reply:
> From what I understand, only the transformed samples are passed to flowMC and also output at the end as a data product. However, in some cases, e.g., using the extrinsic transform, one would also like to keep the original parameter value to avoid the need for a backward transform.
> 
> Moreover, in some special transforms, e.g., the folding transform, by keeping the original value, one does not need to conduct the whole unfolding process. That on its own is quite expensive.
> 
> Overall, it would be better to have an additional list of parameters to keep, or by default, keep everything, as to aid the user.

Transform is a function, as long as one does not overwrite the original variables and no error during the transform, the original variable is preserved. If the user needs any original parameters, I think the can just keep the variables before the transform.

Reply:
> Moreover, how would it interface with the flowMC NF proposal? I would expect that to only be proposing new points in the transformed space. Does it mean that an inverse transform is always needed, either for data product or likelihood calculation?

flowMC only interface with the pre-transform space, so the inverse transform is never required. This is because Jim posterior function is called in the sampling space, which is the pre-transformed space. The current design is we prohibit evaluating log_prob in the transformed space, since that requires an inverse transform in the general setting, and the inverse transform is not guaranteed to exist in general. 

Interfacing with the data product or likelihood calculation only requires forward transform

Reply:
But if the flowMC only interface with the pre-transform space, how could we leverage it with those better extrinsic parameterization?

For example, for distance d, I want to sample in a better transformed version dhat to avoid non-linear correlation.

Reply:
By pre-transform space, I meant the Logistical distribution space, which is the basis of an unconstrained parameter space. Ideally, all of the variables are ultimately traced from a simple distribution, which is currently chosen as a logistics distribution. If the transforms are written in the right way, the distribution in the unconstrained space should be relatively simple. The gradient of the transform should also be encapsulated in the posterior while the transform is executed.

The current philosophy now follows something along the line of https://mc-stan.org/docs/2_18/stan-users-guide/reparameterization-section.html